starlite/datastructures/headers.py

```diff
@@ -1,15 +1,182 @@
+
+    def add(self, name: str, value: str) -> None:
+        """Add a header to the scope keeping duplicates."""
+        self.headers.append((name.lower().encode("latin-1"), value.encode("latin-1")))
+
+    def getall(self, name: str, default: Optional[List[str]] = None) -> List[str]:
+        """Get all values of a header.
+
+        Args:
+            name: Header name
+            default: Default value to return if `name` is not found
+
+        Returns:
+            A list of strings
+
+        Raises:
+            `KeyError` if no header for `name` was found and `default` is not given
+        """
+        name = name.lower()
+        values = [
+            header_value.decode("latin-1")
+            for header_name, header_value in self.headers
+            if header_name.decode("latin-1").lower() == name
+        ]
+        if not values:
+            if default:
+                return default
+            raise KeyError
+        return values
+
+    def extend_header_value(self, name: str, value: str) -> None:
+        """Extend a multivalued header (that is, a header that can take a comma
+        separated list). If the header previously did not exist, it will be
+        added.
+
+        Args:
+            name: Header name
+            value: Header value to add
+
+        Returns:
+            None
+        """
+        existing = self.get(name)
+        if existing is not None:
+            value = ", ".join([*existing.split(","), value])
+        self[name] = value
+
+    def __getitem__(self, name: str) -> str:
+        """Get the first header matching `name`"""
+        name = name.lower()
+        for header in self.headers:
+            if header[0].decode("latin-1").lower() == name:
+                return header[1].decode("latin-1")
+        raise KeyError
+
+    def _find_indices(self, name: str) -> List[int]:
+        name = name.lower()
+        return [i for i, (name_, _) in enumerate(self.headers) if name_.decode("latin-1").lower() == name]
+
+    def __setitem__(self, name: str, value: str) -> None:
+        """Set a header in the scope, overwriting duplicates."""
+        name_encoded = name.lower().encode("latin-1")
+        value_encoded = value.encode("latin-1")
+        indices = self._find_indices(name)
+        if not indices:
+            self.headers.append((name_encoded, value_encoded))
+        else:
+            for i in indices[1:]:
+                del self.headers[i]
+            self.headers[indices[0]] = (name_encoded, value_encoded)
+
+    def __delitem__(self, name: str) -> None:
+        """Delete all headers matching `name`"""
+        indices = self._find_indices(name)
+        for i in indices[::-1]:
+            del self.headers[i]
+
+    def __len__(self) -> int:
+        return len(self.headers)
+
+    def __iter__(self) -> Iterator[str]:
+        """iter over header names including duplicates."""
+        return iter(h[0].decode("utf-8") for h in self.headers)
```

Comment:
I am puzzled to see here `decode("utf-8")` because header names are encoded to `latin-1` everywhere

Reply:
Oh, good catch. That must have been an autocomplete I overlooked (from my brain or IDE :D)